bin/varnishd/cache/cache_req_body.c

```diff
@@ -43,6 +43,47 @@
 #include "storage/storage.h"
 #include "hash/hash_slinger.h"
 
+/*----------------------------------------------------------------------
+ * Check and potentially update req framing headers.
+ */
+
+static ssize_t
+vrb_cached(struct req *req, ssize_t req_bodybytes)
+{
+	ssize_t l0, l;
+
+	CHECK_OBJ_NOTNULL(req, REQ_MAGIC);
+	AZ(req->req_body_cached);
+	assert(req_bodybytes >= 0);
+
+	l0 = http_GetContentLength(req->http0);
+	l = http_GetContentLength(req->http);
+
+	assert(req->req_body_status->avail > 0);
+	if (req->req_body_status->length_known) {
+		assert(req_bodybytes == l0);
+		assert(req_bodybytes == l);
```

Comment:
As `req.http.Content-Length` can be modified from VCL, I think we should rather either fix the header or fail the request rather than panicking.

Reply:
You have my sympathies if you are saying this from a feeling of frustration.
But I actually think we should not not have any panics which can be triggered from VCL.

Also I wanted to take note of an idea from bugwash: For the rollback case, can we make the code DTRT even with the unmodified headers?

Reply:
> You have my sympathies if you are saying this from a feeling of frustration.
> But I actually think we should not not have any panics which can be triggered from VCL.

Actually no, I'm genuinely thinking that if we leave dragons in VCL, we can't afford a gentle failure here. If you both mess with framing headers and request body caching things are going horribly wrong and should blow up.

> Also I wanted to take note of an idea from bugwash: For the rollback case, can we make the code DTRT even with the unmodified headers?

One thought I didn't discuss during bugwash is that body status shouldn't be the responsibility of the HTTP connection. We miss some separation of concerns in `struct http_conn` that was OK when HTTP/1 was the only game in town. There are more things to disentangle.

Reply:
In the course of my [continuation of this work](https://github.com/varnishcache/varnish-cache/pull/3798#issuecomment-1148576467) I replaced the code with [just setting the respective headers](https://github.com/nigoroll/varnish-cache/commit/c371d92c45ac3ff530cca7301c9721adc408248e) because, with filters on the request body (but before caching), we can not make any assumptions on the relation between the content-length before/after caching.